external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DeterministicKafkaInputDStream.scala

```diff
@@ -0,0 +1,123 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.kafka
+
+import scala.annotation.tailrec
+import scala.reflect.{classTag, ClassTag}
+
+import kafka.common.TopicAndPartition
+import kafka.message.MessageAndMetadata
+import kafka.serializer.Decoder
+
+import org.apache.spark.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.rdd.kafka.{KafkaCluster, KafkaRDD}
+import org.apache.spark.streaming.{StreamingContext, Time}
+import org.apache.spark.streaming.dstream._
+
+/** A stream of {@link org.apache.spark.rdd.kafka.KafkaRDD} where
+  * each given Kafka topic/partition corresponds to an RDD partition.
+  * The spark configuration spark.streaming.receiver.maxRate gives the maximum number of messages
+  * per second that each '''partition''' will accept.
+  * Starting offsets are specified in advance,
+  * and this DStream is not responsible for committing offsets,
+  * so that you can control exactly-once semantics.
+  * For an easy interface to Kafka-managed offsets,
+  *  see {@link org.apache.spark.rdd.kafka.KafkaCluster}
+  * @param kafkaParams Kafka <a href="http://kafka.apache.org/documentation.html#configuration">
+  * configuration parameters</a>.
+  *   Requires "metadata.broker.list" or "bootstrap.servers" to be set with Kafka broker(s),
+  *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
+  * @param fromOffsets per-topic/partition Kafka offsets defining the (inclusive)
+  *  starting point of the stream
+  * @param messageHandler function for translating each message into the desired type
+  * @param maxRetries maximum number of times in a row to retry getting leaders' offsets
+  */
+class DeterministicKafkaInputDStream[
+  K: ClassTag,
+  V: ClassTag,
+  U <: Decoder[_]: ClassTag,
+  T <: Decoder[_]: ClassTag,
+  R: ClassTag](
+    @transient ssc_ : StreamingContext,
+    val kafkaParams: Map[String, String],
+    val fromOffsets: Map[TopicAndPartition, Long],
+    messageHandler: MessageAndMetadata[K, V] => R,
+    maxRetries: Int = 1
+) extends InputDStream[R](ssc_) with Logging {
+
+  private val kc = new KafkaCluster(kafkaParams)
+
+  private val maxMessagesPerPartition: Option[Long] = {
+    val ratePerSec = ssc.sparkContext.getConf.getInt("spark.streaming.receiver.maxRate", 0)
```

Comment:
I'm not sure specifically what you mean by "window operations that require
past data which needs to be pulled from Kafka every time". The current
KafkaRDD code has a log every time compute() is called on the rdd to pull
data from kafka, and for a job with a window operation, I only see compute
called once for a given offset range, not repeatedly every time.

Regarding the bigger question of how this approach stacks up to the two
existing approaches... they're all different approaches with different
tradeoffs, I don't think one has to win.  I'd still have a use for the
original receiver based class (not the WAL one), especially if SPARK-3146
or SPARK-4960 ever get merged.

On Sat, Jan 3, 2015 at 8:57 PM, Tathagata Das <notifications@github.com>
wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DeterministicKafkaInputDStream.scala
> <https://github.com/apache/spark/pull/3798#discussion-diff-22436219>:
>
> > +  K: ClassTag,
> > +  V: ClassTag,
> > +  U <: Decoder[_]: ClassTag,
> > +  T <: Decoder[_]: ClassTag,
> > +  R: ClassTag](
> > +    @transient ssc_ : StreamingContext,
> > +    val kafkaParams: Map[String, String],
> > +    val fromOffsets: Map[TopicAndPartition, Long],
> > +    messageHandler: MessageAndMetadata[K, V] => R,
> > +    maxRetries: Int = 1
> > +) extends InputDStream[R](ssc_) with Logging {
> > +
> > +  private val kc = new KafkaCluster(kafkaParams)
> > +
> > +  private val maxMessagesPerPartition: Option[Long] = {
> > +    val ratePerSec = ssc.sparkContext.getConf.getInt("spark.streaming.receiver.maxRate", 0)
>
> Yeah, I get it. However, this still has a different performance
> characteristics than receiver based sources. Sometimes it is better (as you
> are "always" pulling data in parallel across cluster, instead of the
> default 1 receiver), sometimes it is worse (window operations that require
> past data which needs to be pulled from Kafka every time). For this method
> to be viable alternative to the existing, we ideally have to make sure that
> the performance characteristics of this method is >= performance of the
> existing method under all situations. Then using this method will be
> justified even if the behavior is different.
>
> On that note, here is an idea of what we can do. We can store the data
> pulled from the Kafka as blocks in the BlockManager, so that subsequent
> accesses to the data (due to window or stateful ops) can be faster. One way
> to do this is to implement a KafkaBackedBlockRDD. This is similar to
> WriteAheadLogBackedBlockRDD which has the logic to either read from
> BlockManager if block is present, or reload the data from the WriteAheadLog
> based on file segment info. KafkaBackedBlockRDD can be similar, either
> read from BlockManager, or load it from Kafka based on offsets.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r22436219>.
>